src/vcl/utils.rs

```diff
@@ -1,7 +1,66 @@
+use std::ffi::{c_char, CStr};
+use std::slice::from_raw_parts;
+use std::str::from_utf8;
+
 use crate::ffi;
 use crate::ffi::{vfp_entry, vrt_ctx};
 use crate::vcl::backend::VCLBackendPtr;
 
+impl ffi::txt {
+    pub fn from_bytes(s: &[u8]) -> Self {
+        Self {
+            b: s.as_ptr().cast::<c_char>(),
+            e: unsafe { s.as_ptr().add(s.len()).cast::<c_char>() },
+        }
+    }
+
+    /// FIXME: This method is only used when calling [`ffi::VSLbt`],
+    /// and current implementation creates a string without a null terminator to pass it in.
+    /// Going forward, we should probably refactor it to avoid extra string allocation.
+    #[allow(clippy::should_implement_trait)]
+    pub fn from_str(s: &str) -> Self {
+        Self::from_bytes(s.as_bytes())
+    }
+
+    pub fn from_cstr(s: &CStr) -> Self {
+        Self::from_bytes(s.to_bytes())
+    }
+
+    /// Convert the `txt` struct to a `&[u8]`.
+    /// FIXME: Consider modifying this to return an empty slice instead of None,
+    ///        Unless we want to explicitly differentiate between empty and null.
```

Comment:
we absolutely want to distinguish between empty and null
```suggestion
```

Reply:
thx, I updated the comment